pkg/admission/pod_eviction.go

```diff
@@ -0,0 +1,464 @@
+
+	// accumulateResult is called for each StatefulSet which is tested
+	accumulateResult func(sts *appsv1.StatefulSet, result *podStatusResult) error
+
+	// validate is called after all the StatefulSets have been tested - this function validates that the PDB will not be breached
+	validate func(maxUnavailable int) error
+
+	// successMessage returns a success message we return in the eviction response
+	successMessage func() string
+}
+
+// plural appends an 's' to the given string if the value is > 1.
+func plural(s string, value int) string {
+	if value > 1 {
+		return fmt.Sprintf("%ss", s)
+	}
+	return s
+}
+
+// pdbMessage creates a message which includes the number of not ready and unknown pods within the given span
+func pdbMessage(result *podStatusResult, span string) string {
+	msg := ""
+	if result.notReady > 0 {
+		// 1 pod not ready
+		msg += fmt.Sprintf("%d %s not ready", result.notReady, plural("pod", result.notReady))
+		if result.unknown > 0 {
+			msg += ", "
+		}
+	}
+	if result.unknown > 0 {
+		// 2 pods unknown
+		msg += fmt.Sprintf("%d %s unknown", result.unknown, plural("pod", result.unknown))
+	}
+
+	// under ingester-zone-a partition 0
+	// under ingester-zone-a
+	if len(span) > 0 {
+		msg += fmt.Sprintf(" under %s", span)
+	}
+	return msg
+}
+
+// a lock used to control finding a specific PDB lock
+var lock = sync.RWMutex{}
+
+// a map of ZPDB name to lock
+var locks = make(map[string]*sync.Mutex, 5)
+
+// findLock returns a Mutex for each unique name, creating the Mutex if required.
+func findLock(name string) *sync.Mutex {
+	lock.RLock()
+	if pdbLock, exists := locks[name]; exists {
+		lock.RUnlock()
+		return pdbLock
+	}
+	lock.RUnlock()
+
+	lock.Lock()
+	defer lock.Unlock()
+
+	if pdbLock, exists := locks[name]; exists {
+		return pdbLock
+	}
+
+	locks[name] = &sync.Mutex{}
+	return locks[name]
+}
+
+// A admissionReviewRequest holds the context of the request we are processing.
+type admissionReviewRequest struct {
+	log    *spanlogger.SpanLogger
+	req    v1.AdmissionReview
+	client k8sClient
+}
+
+// canIgnorePDB returns true for pod conditions that allow the pod to be deleted without checking PDBs.
+// Adapted from https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go
+func canIgnorePDB(pod *corev1.Pod) bool {
+	if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed ||
+		pod.Status.Phase == corev1.PodPending || !pod.DeletionTimestamp.IsZero() {
+		return true
+	}
+	return false
+}
+
+// validate ensures that the AdmissionRequest contains the information we require to process this request.
+func (r *admissionReviewRequest) validate() error {
+	// ignore non-create operations
+	if r.req.Request.Operation != v1.Create {
+		return errors.New("request operation is not create")
+	}
+
+	// ignore create operations other than subresource eviction
+	if r.req.Request.SubResource != eviction {
+		return errors.New(`request SubResource is not eviction`)
+	}
```

Comment:
For both of these cases, it'd be good to include information on what we did get (ie. the operation / sub-resource we got)